concepts/pipeline/data_pipeline_detail.md

```diff
@@ -0,0 +1,429 @@
+
+## Monitoring and Alerting
+
+There are multiple layers of monitoring and alerting.
+
+At a low level, the system is monitored to ensure that it is functioning as expected. This includes things like machine-level resources (network capacity, disk space, available RAM, CPU load) which are typically monitored using [DataDog].
+
+Next, we monitor the "transport" functionality of the system. This includes monitoring incoming submission rates, payload sizes, traffic patterns, schema validation failure rates, and alerting if anomalies are detected. This type of anomaly detection and alerting is handled by [Hindsight].
+
+Once data has been safely ingested and stored, we run some automatic regression detection on all Telemetry [histogram measures] using [Cerberus]. This code looks for changes in the distribution of a measure, and emails probe owners if a significant change is observed.
+
+Production ETL jobs are run via [Airflow], which monitors batch job progress and alerts if there are failures in any job. Self-serve batch jobs run via ATMO also generate alerts upon failure.
+
+Scheduled [Re:dash] queries may also be configured to generate alerts, which is used to monitor the last-mile user facing status of derived datasets. Re:dash may also be used to monitor and alert on high-level characteristics of the data, or really anything you can think of.
+
+## Data Exports
+
+Data is exported from the pipeline to a few other tools and systems. Examples include integration with [Amplitude] for mobile and product analytics, publishing reports and visualizations to the [Mozilla Data Collective], and shipping data to other parts of the Mozilla organization.
+
+There are also a few data sets which are made publicly available, such as the [Firefox Hardware Report].
+
+## Bringing it all together
+
+Finally, here is a more detailed view of the entire platform. Some connections are omitted for clarity.
+
+```mermaid
+graph LR
+ subgraph Data Producers
+  Firefox
+  more_producers[...]
+ end
+ subgraph Storage
+  Landfill
+  warehouse_heka[Heka Data Lake]
+  warehouse_parquet[Parquet Data Lake]
+  warehouse_analysis[Analysis Outputs]
+  PostgreSQL
+  Redshift
+  MySQL
+  hive[Hive] -.- warehouse_parquet
+ end
+ subgraph Stream Processing
+  cep[Hindsight Streaming]
+  dwl[Data Store Loader] --> warehouse_heka
+  dwl --> warehouse_parquet
+  sparkstreaming[Spark Streaming] --> warehouse_parquet
+ end
+ subgraph Ingestion
+  Firefox --> lb[Load Balancer]
+  more_producers --> lb
+  lb --> tee
+  tee --> mozingest
+  mozingest --> kafka
+  mozingest --> Landfill
+  ZooKeeper -.- kafka[Kafka]
+  kafka --> dwl
+  kafka --> cep
+  kafka --> sparkstreaming
+ end
+ subgraph Batch Processing
+  Airflow -.->|spark|tbv[telemetry-batch-view]
+  Airflow -.->|spark|python_mozetl
+  warehouse_heka --> tbv
+  warehouse_parquet --> tbv
+  warehouse_heka --> python_mozetl
+  warehouse_parquet --> python_mozetl
+  tmo_agg[Telemetry Aggregates]
+ end
+ subgraph Visualization
```

Comment:
Can we exclude this subgraph? It doesn't appear to interact with any other component. Maybe the Viz section should be a separate article?

Reply:
I wanted to include more connections in this graph, but without the ability to do any explicit layout in mermaidjs, it quickly turns into a tangled mess.